Normative: Add initializer callback for side effects

littledan opened the pull request:
Many decorators will want to perform a side effect when instantiating
a class, for example:
- To select [[Set]] rather than [[DefineOwnProperty]] semantics for
  defining a field
- To store the field in an entirely different place, neither an ordinary
  property or private field (e.g., MobX stores some fields in a Map)
- To register the instance under construction in some way (e.g., from
  a class decorator). "Instance finishers" have been proposed for this
  purpose, but it's not clear when to run such a finisher. For many use
  cases (e.g., interacting with the DOM), an instance "starter" is just
  as good, due to run-to-completion semantics.

This patch adds a third decorator descriptor kind of "initializer" to register
an initializer to run, interspersed with field initializers.

This PR depends on #163 and #164 

Fixes #44

littledan committed:
Normative: Add `kind: "initializer"` for side effects

Many decorators will want to perform a side effect when instantiating
a class, for example:
- To select [[Set]] rather than [[DefineOwnProperty]] semantics for
  defining a field
- To store the field in an entirely different place, neither an ordinary
  property or private field (e.g., MobX stores some fields in a Map)
- To register the instance under construction in some way (e.g., from
  a class decorator). "Instance finishers" have been proposed for this
  purpose, but it's not clear when to run such a finisher. For many use
  cases (e.g., interacting with the DOM), an instance "starter" is just
  as good, due to run-to-completion semantics.

This patch permits decorators to create elements of the form
  { kind: "initializer", placement: "own", initializer: fn }
to be used purely for their side effect, with evaluation semantics
and ordering identical to fields, and interspersed with them in
evaluation order.

littledan commented:
cc @nicolo-ribaudo @wycats @mweststrate

ljharb reviewed (changes requested):
If key is null and any placement is provided, that looks to me like programmer error. Could we instead throw when a placement is provided unnecessarily, just like we throw in #30?

littledan commented:
@ljharb I agree that error checking in general should be strict where there's no reasonable meaning, but in #57 I think we settled on *not* artificially restricting expressiveness where there is something that could make sense. That's the principle that I'm trying to follow here.

ljharb commented:
How does providing a placement for “no property” make sense?

littledan commented:
@ljharb It affects when it runs and what the receiver is. The typical usage will be to pass `"own"`, so it's called on instances during instance construction. Not providing a placement is an error in this patch.

rdking commented:
Isn't this trying entirely too hard to avoid the constructor? This is very quickly approaching the "magic" annotations of Java. Adding repeatable nonsense to the list of properties just to trigger side effects can:
1. make tooling more difficult to write since it would have to understand `null` as a "special" property name where currently `{[null]:1}` is perfectly valid (albeit weird) code.
2. confuse developers since multiple properties of the same name appear to be being defined on the object
3. be easier done and easier to understand just by using the tried and true `constructor`.

ljharb commented:
@littledan gotcha, that makes sense then - thanks.

littledan commented:
@rdking 

1. For the `{[null]:1}` case: Note that the system will have already turned it into `"null"` by the time it reaches the decorator, so no loss in expressivity.
2. This patch doesn't affect that property; you can have duplicate (ordinary) field and method declarations without this patch (but it's not really useful, given how coalescing works)
3. The purpose here is to help frameworks or similar things insert logic into the path. Lots of people have run into similar problems, which led to decorators in general, and in particular this PR.

jridgewell commented:
I like the idea, but it seems a bit strange to have a field name provided to the decorator for it to just throw it away. How hard would it be to have a decorator without a name, just a init?

```js
class Ex {
  @dec x = 1
  // no init
  @sideEffect ;

  // or some init
  @sideEffect this.x = 2; // Option A
  @sideEffect { this.x = 2 }; // Option B
}
```

rdking commented:
@jridgewell That makes more sense visually, but what would the decorator receive? Is the side effect independent of the field it's supposed to act against?

rdking commented:
Better question, if the decorator is just supposed to trigger some side-effect without regard to any particular field, then why isn't it a class-level decorator?

littledan commented:
@jridgewell I am not sure if we need that syntax extension. The hope here is that these side effects will be used by semantically field-like declarations, or added by a class decorator. Do you have a use case in mind for those other forms?

jridgewell commented:
I guess not, it was just my first reaction.

littledan commented:
I probably didn't explain the use cases very clearly in the above commit message. As an example, you could make a decorator to give a field declaration `[[Set]]` semantics, so the following would work:

```js
class MyComponent extends HTMLElement {
  @set onclick = (event) => { /* ... */ }
}
```

As the following:

```js
function set(desc) {
  assert(desc.placement === "own");
  assert(desc.kind === "field");
  assert((typeof desc.key)[0] === "s");  // symbol or string
  assert(desc.descriptor.configurable);
  assert(desc.descriptor.writable);
  assert(desc.descriptor.enumerable);
  return {key: null, placement: "own", descriptor: { },
          initializer() { this[desc.key] = desc.initializer.call(this) } };
}
```

nicolo-ribaudo commented:
Instead of `key: null`, I think that `kind: "initializer"` would be better. An `"initializer"` must have the `placement` and `initializer` fields, and it throws if it has `key` or `descriptor`. Reusing existing `kind: "field"` to declare something that _is not_ a field (since it doesn't get defined anywhere) seems like unnecessary overloading.

littledan commented:
@nicolo-ribaudo That suggestion sounds totally reasonable to me. Want to write it up in a replacement PR?

nicolo-ribaudo commented:
I can't probably do it this week since I don't have my PC, but if no one does it first I hope I will be able to do it next week.

littledan commented:
Updated to include `kind: "initializer"` as recommended by @mweststrate and @nicolo-ribaudo . Documentation in the explainer still needed (anyone interested?). Reviews welcome!

nicolo-ribaudo commented on spec.html:
```diff
@@ -784,23 +817,35 @@ <h1>ToElementDescriptor ( _elementObject_ )</h1>
         1. Let _key_ be ? Get(_elementObject_, `"key"`).
         1. If _key_ has a [[PrivateName]] internal slot, set _key_ to _key_.[[PrivateName]].
+        1. If _kind_ is `"initializer"`,
+          1. If _key_ is not *undefined*, throw a *TypeError* exception.
```
Maybe we should throw _before_ converting `key` to a valid property key?

littledan replied:
Yes of course; also this broke the if/else conditional!

nicolo-ribaudo commented on spec.html:
```diff
@@ -689,7 +720,7 @@ <h1>DecorateConstructor ( _elements_, _decorators_ )</h1>
           1. If _elementsAndFinisher_.[[Elements]] is not *undefined*,
             1. Set _elements_ to the concatenation of _elementsAndFinisher_.[[Elements]] and _privateElements_.
-            1. If there are two class elements _a_ and _b_ in _elements_ such that _a_.[[Key]] is _b_.[[Key]] and _a_.[[Placement]] is _b_.[[Placement]], throw a *TypeError* exception.
+            1. If there are two class elements _a_ and _b_ in _elements_ such that _a_.[[Kind]] is not `"initializer"` and _b_.[[Kind]] is not `"initializer"`, _a_.[[Key]] is _b_.[[Key]], and _a_.[[Placement]] is _b_.[[Placement]], throw a *TypeError* exception.
```
Can this be made more readable? (I don't know if it valid from a spec-grammar point of view)

```
1. If there are two class elements _a_ and _b_ in _elements_ such that all the following conditions are true:
  1. _a_.[[Kind]] is not `"initializer"`
  1. _b_.[[Kind]] is not `"initializer"`
  1. _a_.[[Key]] is _b_.[[Key]]
  1. _a_.[[Placement]] is _b_.[[Placement]]
1. Then throw a *TypeError* exception.
```

littledan replied:
Good idea, done

nicolo-ribaudo commented on spec.html:
```diff
@@ -784,23 +817,35 @@ <h1>ToElementDescriptor ( _elementObject_ )</h1>
         1. Let _elements_ be ? Get(_elementObject_, `"elements"`).
         1. If _elements_ is not *undefined*, throw a *TypeError* exception.
-        1. If _kind_ not `"field"`,
+        1. If _kind_ not `"field"` or `"initializer"`,
```
Typo: "If kind **is** not"

littledan replied:
Done

nicolo-ribaudo commented on spec.html:
```diff
@@ -188,6 +188,16 @@ <h1>The ElementDescriptor Specification Type</h1>
+            <ul>
+              <li>_element_.[[Key]] absent.</li>
+              <li>_element_.[[Descriptor]] is absent.</li>
+              <li>_element_.[[Placement]] is `"own"`.</li>
```
We might consider allowing "static" as the decorator way of doing https://github.com/tc39/proposal-class-static-block

ljharb replied:
I agree; static should be allowed here since they also can have initializers

littledan replied:
I don't see this as a way towards a static block. This is just a way that a *decorator* can add extra code to run. But decorators are already able to run code at the end of when a class is executing, in the finisher callback. We could allow this out of consistency (it would have slightly different timing from finishers), but I don't see a use case.

ljharb replied:
A use case is the same as for "own" - being able to invoke a setter on a superclass (in this case, on a superclass's constructor that had, say, `static set foo(v) {}`)

littledan replied:
@ljharb What would be the problem with using a finisher for that?

ljharb replied:
Certainly you could use a finisher for that - but then you'd be forced to use a class-level decorator to change the semantics of a specific field, so instead of:
```jsx
class C {
  @useSet static a = x;
  static b = y;
  @useSet static c = z;
}
```
you'd need to have:
```jsx
@useSet('a')
@useSet('c')
class C {
  static a = x;
  static b = y;
  static c = z;
}
```
or:
```jsx
@useSet('a', 'c')
class C {
  static a = x;
  static b = y;
  static c = z;
}
```
both of which are decidedly less ergonomic and elegant.

littledan replied:
I don't see why that would be the case. You can use a finisher on a static field declaration too.

Lodin replied:
@ljharb, wait, but you **can** use `finisher` with a field or a method decorator. It is not required to use class-level decorators to activate finishers.

ljharb replied:
ahh, that wasn't clear to me. In that case - instead of this "initializer" approach, is there a reason not to allow finishers on "own" fields, to provide an "instance finisher"?

Lodin replied:
@ljharb yeah, I came to the same idea in my [comment](https://github.com/tc39/proposal-decorators/pull/165#issuecomment-435520319) below :)

nicolo-ribaudo reviewed (commented):
1. I don't remember its name right now, but there is a function which checks that there aren't duplicated keys which probably should be updated to handle ElementDescriptors without [[Key]].
2. Maybe we should throw if the `initializer` property of a `{ kind: "initializer" }` element is `undefined`, since it would be almost certainly a programming error?

littledan committed:
Fix most of the errors from review

littledan committed:
Normative: Throw an exception on a missing initializer

Lodin commented:
@littledan I have a concern about `finisher` and initializer field. They kinda duplicate each other. E.g. if initializer fields can be `static` it almost removes profit finisher brings. If they cannot, it looks strange: why they cannot do it?

Then I came to a thought: why not to do the `constructor` callback that works exactly as finisher, but during class instantiation? It would reduce not only amount of code that is necessary to describe the logic, but would also be clearer and obvious to use. Or am I missing something?

ljharb commented on spec.html:
```diff
@@ -835,7 +833,10 @@ <h1>ToElementDescriptor ( _elementObject_ )</h1>
+        1. If _kind_ is `"initializer"` and _initializer_ is *undefined*, throw a *TypeError* exception
         1. Let _elements_ be ? Get(_elementObject_, `"elements"`).
+        1. If _kind_ is `"initializer"`,
+          1. If _key_ is not *undefined*, throw a *TypeError* exception.
```
shouldn't this check happen right after `Let _key_ be ? Get(_elementObject_, `"key"`)`?

littledan replied:
I previously tried to sort the validations towards the end, but I can do them mixed inline if you prefer.

ljharb replied:
I feel like in general we strive to have validations occur as soon as possible, to avoid unnecessary observable behavior.

littledan commented:
@Lodin I agree that there's overlap between initializers and fields; that's why I started off this PR by saying it was a field with a null key. I also agree that static initializers are not extremely interesting; that's why this patch omits them. A `constructor` callback could be a fine way to add *additional* behavior to a descriptor that's already there, but if you want the *only* behavior to be running the callback (as in the `@set` example), I'm not sure how we'd make it work.

Lodin commented:
@littledan Oh I see. My bad, started to read, caught the idea and immediately wrote it. Thanks for explanation.

Well, then, maybe decorators don't need `finisher`s? Initializer fields could be used instead.

littledan commented:
Finishers are something different, for when the class itself is done being created. They are useful for instance elements as well, e.g., to attach runtime metadata.

Lodin commented:
@littledan I mean, it could be solved this way:
```javascript
const customElement = name => ({
  kind,
  elements,
}) => {
  assert(kind === 'class');
  
  return {
    kind,
    elements: [
      ...elements,
      {
        descriptor: {},
        key: null,
        placement: 'static',
        initializer(target) {
          customElements.define(name, target);
        },
      },
    ],
  };
}
```
All `static` initializer fields could be called at the time `finisher` is called now, after the class creation is finished.

Kinda verbose, but now we removed overlapping and have only one type of initializers.

littledan commented:
Finishers all run after static field initializers, not interspersed, in case, e.g., you want to freeze the whole class.

I am not sure how far we will get looking for a completely minimal calculus here. Most of this is syntactic sugar and somehow redundant anyway. It is not the end of the world if there are two similar ways to do things.

ljharb commented:
Is there a reason we wouldn't want an instance finisher?

Lodin commented:
@littledan I see it now. Thanks!

@ljharb You cannot implement [this case](https://github.com/tc39/proposal-decorators/pull/165#issuecomment-434434466) with just an instance finisher.

ljharb commented:
@Lodin if it ran per-instance, it seems like you could?

littledan commented:
Lots of people have brought up instance finishers, but I still haven't seen a complete proposal for what the semantics would be. I worry that adding an extra step to instantiation could be difficult without performance regressions, but this is a vague worry since I don't know what the semantics are.

In any case, instance elements still need the ability to run class-level finishers, e.g., to attach metadata.

ljharb commented:
If static finishers are sufficient to handle side-effecting decorators on static fields, then it seems instance finishers should be used for instance fields as well - but if instance fields need side-effecting decorators in the way the PR introduces, then I’m not sure why static fields wouldn’t also.

littledan commented:
I still don't know what people want out of instance finishers. If you tell me, maybe I can think about how to replace these with those. I'm fine with adding static initializers, but I still don't see a use case, whereas I also am hesitant about instance finishers, as explained in https://github.com/tc39/proposal-decorators/pull/165#issuecomment-435576037

mweststrate commented:
Thanks @littledan! As far as I can see this matches all my existing use cases :)

littledan commented:
To @ljharb 's point on the overlap between finishers and initializers: This patch ignores the return value of initializers, but a finisher which returns non-undefined can replace the constructor. The other issue is a subtle one about ordering: All the finishers run after all the field declarations. If we were to try to have initializers subsume finishers, as @ljharb seems to be suggesting above, we'd make the following changes:
- Permit both instance and static initializers (and prototype, why not?)
- Run the initializers after the field declarations (so it's a three-step process: methods, fields, initializers)
- Use the return value for the initializer to replace the object under construction (whether it's the instance or the constructor--or maybe we would ban instance and prototype initializers returning non-undefined?)

(Should we rename initializer to finisher? It's a bit confusing to juggle the term initializer for what this PR introduces and initializer for what a field has.)

rbuckton commented:
>  As an example, you could make a decorator to give a field declaration [[Set]] semantics, so the following would work: […]

I still argue that a field should have \[\[Set]] semantics by default. That said, do we really need this if you can do constructor replacement in a finisher?

ljharb commented:
Regardless of whether Set or Define was the default, I think you'd want to be able to use a decorator on a single field to have the other semantic, without having to replace the entire constructor and deal with the `.constructor` and prototype chain difficulties there in.

littledan commented:
Let's keep the Set vs Define discussion in the class fields repo.

@rbuckton I don't think you would even need constructor replacement for it, just the ability to perform a side-effecting operation (which you would get either from finishers or static initializers). 

The above post is about, *if* we want to try to replace finishers with initializers (as @ljharb suggested above), then how would we do it.

Do you see any part of the above plan that wouldn't provide a replacement for finalizers while enabling the kind of decorator described in the commit message? Would it be too complicated, or unergonomic?

rbuckton commented:
I am very concerned this will further muddy the waters for the decorators proposal. I think we would be better served to consider this as a follow-on proposal rather than adopt it as part of the current proposal.

In addition I feel it hijacks the semantics of `initializer`, since for the case where `key` is not `null`, the initializer return value is important, while when `key` is `null`, the `initializer` return value is ignored. It very-much feels like we are mixing metaphors and that this could be a source of confusion.

If we were to add this, I would be more willing to accept something like `finisher` for instances that runs when the current class constructor is evaluated, though not the previously proposed "instance finisher" semantics that waited to evaluate instance finishers until after _all_ constructors (including subclasses) were run.

littledan commented:
@rbuckton Are you talking about this PR or the plan in https://github.com/tc39/proposal-decorators/pull/165#issuecomment-435690673 ? The later comment would be like your last paragraph, except the "instance finisher" would run just after all the field initializers and before the constructor body.

nicolo-ribaudo commented:
How could an instance finisher work _in a nice way_? I'm trying to understand it using static fields/class finishers, but I'm struggling to implement even simple thigns (replace [[Define]] with [[Set]]): how can I _replace_ a field with a finisher?

This works, but it puts the decorator and the field in different places:
```js
@useSetOn("foo")
class A extends Base {
  static foo = 2;
}

function useSetOn(name) { return _useSetOn.bind(null, name); }
function _useSetOn(name, descriptor) {
  let index = descriptor.elements.findIndex(el => el.key === name);
  let { initializer } = descriptor.elements[index];
  descriptor.elements.splice(index, 1);
  descriptor.finisher = Class => {
    Class.foo = initializer.call(Class);
  };
}
```
[demo](https://deploy-preview-1816--babel.netlify.com/repl#?babili=false&browsers=&build=master&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEBCkFNoG8BQ1oQC5iwS2EwS2gDMB7cgCgDcBKFdDaMEBAJyyoCILyiSWft2gBqaPQDcTAL6o5qAAIBXCAgDKxAPIA7Hn251UoSDACC0BAA8sCHQBMY8NYwzZcBMpWgBeaACZpBVJlHWB8ch1oVQ1tPR0wAFsEBmRodmJldiiAfRjNLF0AOgAjPAcqHWUQEAAaaATkukloORCwiNz8uMqkhHr7BAhgdjwAByF2VKY2EnLBq19oQeHRifJ2IoQ2ZJ0sCCLSeYBJB2sqbd8APksQIoBrBABPXx8_RpTpDFnocrx8Vh4ABeHCWKxG40mWx2dn2AG15tYALpFP4AkDAjhfZZDCHrTbbBC7fZFCBjDHABBURFWeoARmaTHBayhRx0eAgAAtQX4AMLgKDXVzMfmmQ7ePxovCAkGbYCsEBUUVQRkYGRBIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=stage-2&prettier=false&targets=&version=7.1.2&envVersion=)

This doesn't work (it throws):
```js
class A extends Base {
  @useSet static foo = 2;
}

function useSet(descriptor) {
  let { initializer, key } = descriptor;
  return {
    finisher(Class) {
      Class[key] = initializer.call(Class);
    }
  };
}
```
[demo](https://deploy-preview-1816--babel.netlify.com/repl#?babili=false&browsers=&build=master&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEBCkFNoG8BQ1oQC5iwS2EwS2gDMB7cgCgDcBKFdDaMEBAJyyoCILyiSWft2gBqaPQDcTAL6o5qUJBgBBaAgAeWBADsAJjHgQkaDAAEArsYDKxTDnyE-0ALzQATNIWkLO4PnIdaCsEWy49BAhgdjwAByF2BlNoNhJkaDwdPHxWPAAvDgAaaABrBABPaBlXaAiomPjydmkMdmILdiDkjFJMvAgACw4qAGFwKCSmZmgx5QBtMvKAXRq-nJB8jgA6YFYQUfGIOhbmOQwZLyA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=stage-2&prettier=false&targets=&version=7.1.2&envVersion=)

---

Btw, with finishers instead of initializers this would never work:
```js
class A {
  @useSet foo = 2;
  bar = this.foo;
}
```

ljharb commented:
I was envisioning a field decorator able to run an instance finisher as well, to avoid that very problem.

nicolo-ribaudo commented:
I used class finishers with static fields, but if we had instance finishers I could ask the same question for instance field.

littledan commented:
@nicolo-ribaudo This is a good point. It seems like I was being too optimistic about one thing subsuming the other. Maybe we should just stick with the `own`-only initializers proposed in this PR.

ljharb commented:
If finishers aren’t a solution, then I’m still not sure why this PR wouldn’t cover both own and static.

nicolo-ribaudo commented:
A possible fix to my point would be to make finishers be declared like normal element descriptors, similarly to the `initializer`s proposed in this PR. The decorator would then become
```js
function useSet(descriptor) {
  let { initializer, key } = descriptor;
  return {
    kind: "finisher",
    placement: "static",
    finisher(Class) {
      Class[key] = initializer.call(Class);
    }
  };
}
```

However, if I had to choose I'd still prefer `initializer`s, since they allow me to keep the correct ordering in cases like
```js
class A {
  @useSet foo = 2;
  bar = this.foo;
}
```

nicolo-ribaudo commented:
@rbuckton

> In addition I feel it hijacks the semantics of `initializer`, since for the case where `key` is not `null`, the initializer return value is important, while when `key` is `null`, the `initializer` return value is ignored. It very-much feels like we are mixing metaphors and that this could be a source of confusion.

If `key` is `null`, the initializer's return value isn't discarded but it is used to initialize a `"null"` property:
```js
class A {
  @dec foo = 2;
}

function dec(desc) {
  return {
    ...desc,
    key: null
  }
}

(new A)[null]; // 2
```

littledan committed:
Normative: Permit prototype and static initializers

Also move error checks in ToElementDescriptor to happen ASAP

littledan commented:
@ljharb @nicolo-ribaudo Oh, right. Thanks for bearing with me. The most recent commit permits initializers to be static (or prototype, since why not), and doesn't make any of the other changes mentioned, since we've deduced upthread that they don't make sense. Thoughts?

littledan commented:
One question remains: Should we look at the return value of the initializer callback, and throw an exception if it's not undefined? This could help avoid programmer errors of the kind @rbuckton may be thinking of, and it could make the non-undefined return value a possible future extension point.

nicolo-ribaudo commented:
> One question remains: Should we look at the return value of the initializer callback, and throw an exception if it's not undefined? This could help avoid programmer errors of the kind @rbuckton may be thinking of, and it could make the non-undefined return value a possible future extension point.

:+1: Decorators, even being as powerful as possible, should be as strict as possible to avoid as many potential errors as possible.

ljharb commented on spec.html:
```diff
@@ -563,7 +617,7 @@ <h1>Element Descriptors</h1>
       <pre><code class=typescript>
         interface ElementDescriptor {
-          kind: "method" or "field"
+          kind: "method", "initializer" or "field"
```
```suggestion
          kind: "method", "initializer", or "field"
```

littledan replied:
Done

ljharb commented on spec.html:
```diff
@@ -689,7 +743,12 @@ <h1>DecorateConstructor ( _elements_, _decorators_ )</h1>
           1. If _elementsAndFinisher_.[[Elements]] is not *undefined*,
             1. Set _elements_ to the concatenation of _elementsAndFinisher_.[[Elements]] and _privateElements_.
-            1. If there are two class elements _a_ and _b_ in _elements_ such that _a_.[[Key]] is _b_.[[Key]] and _a_.[[Placement]] is _b_.[[Placement]], throw a *TypeError* exception.
+            1. If there are two class elements _a_ and _b_ in _elements_ such that all of the following are true:
```
This reads a bit awkwardly; maybe it should be in an abstract operation?

littledan replied:
This is following @nicolo-ribaudo 's suggestion above. I think the factoring might be awkward too; let's work out these editorial issues during Stage 3.

ljharb commented on spec.html:
```diff
@@ -781,26 +842,38 @@ <h1>ToElementDescriptor ( _elementObject_ )</h1>
         1. Assert: _elementObject_ is an ECMAScript language value.
         1. Let _kind_ be ? ToString(? Get(_elementObject_, `"kind"`)).
-        1. If _kind_ is not one of `"method"` or `"field"`, throw a *TypeError* exception.
+        1. If _kind_ is not one of `"initializer"`, `"method"` or `"field"`, throw a *TypeError* exception.
```
```suggestion
        1. If _kind_ is not one of `"initializer"`, `"method"`, or `"field"`, throw a *TypeError* exception.
```

littledan replied:
Done

ljharb commented on spec.html:
```diff
@@ -781,26 +842,38 @@ <h1>ToElementDescriptor ( _elementObject_ )</h1>
+        1. If _kind_ is `"method"`,
+          1. If _initializer_ is not *undefined*, throw a *TypeError* exception.
+        1. Let _elements_ be ? Get(_elementObject_, `"elements"`).
+        1. If _elements_ is not *undefined*, throw a *TypeError* exception.
```
Should this use a HasProperty, instead of a Get, since it’s just checking existence? Or is the intention to allow an explicit undefined property - and if so, that seems like it’d be an error, so why allow it?

littledan replied:
The Get matches the use of Get when actually using the elements in a class decorator, per POLS.

littledan committed:
Normative: Throw a TypeError if an initializer returns undefined

littledan committed:
Documentation for initializers

littledan committed:
Add Oxford commas

littledan commented:
Any further concerns with this patch? @rbuckton It's unclear to me whether we've addressed your concerns.

In my opinion, it's important to get this change in Decorators v1, because I don't want an idiom of "throwaway fields" (used for their side effect) to emerge. I believe it would be very difficult for implementations to optimize away the storage space used in throwaway fields, so I don't want to encourage it in the ecosystem. These have been recommended in various issue threads, and the need for a side effect in this path has been known for more than a year.

littledan commented:
OK, given the all-around positive reviews, merging this PR. We'll review it in the November 2018 TC39 meeting, and revert or revisit if necessary. Thanks for the help, everyone!
